### ISSUE TITLE
Provide `spawn` macro

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio#1180]
Another take on #626

Remove `spawn` function and replace it with a `spawn!` macro. The macro would return a handle to the spawned task. The handle would be annotated with `must_use`. The user would be expected to `await` on the return handled or to explicitly call `handle.detach()` (naming tbd).

Additionally, `handle.cancel()` would explicitly force the task to cancel.

Alternatively, the default could be detached (no `must_use` annotation) and the user would need to call a fn like `join`: `my_handle.join().await`. This would line up more with `std` threads but may not be ideal from an "avoiding errors" point of view.

## Questions

- If `cancel`, `await`, or `detach` are not called, should `SpawnHandle::drop` panic?

## Macro

The reason to use a macro instead of a function would be to support passing in additional arguments to name the spawned task. This would tie in with tokio-trace in that it the task name / attributes would be exposed via the tracing system. Then, tools like tokio console (work in progress) could inspect this to track task execution:

```rust
let var = get_var();

let handle = spawn!(async {
    do_something(var).await
}, "my task", var);

let res = handle.await;
```

Depends on #887.

cc @hawkw for tokio-trace ref.

**Comment by leodasvacas** [comment]
Tracing information is not essential, so I think the builder pattern used to construct a named `std::thread` would work fine here.

**Comment by jhpratt** [comment]
Any intent to implement on this? I'm specifically interested in the ability to get a handle to a `Future` that is also spawned on the executor. It would be useful for SergioBenitez/Rocket#1097, where we're looking into returning a `Future` that outputs the same value as the spawned one.

I can look into doing this (or even just a `spawn_with_handle` method like [`futures-preview`](https://docs.rs/futures-preview/0.3.0-alpha.18/futures/task/trait.SpawnExt.html#method.spawn_with_handle)) if someone can point me in the right direction.

**Comment by carllerche** [comment]
Yes! This will be in the final 100%. For now spawn with handle works!

**Comment by carllerche** [comment]
Getting a handle is done on master.
